main.py

```diff
@@ -7,9 +7,151 @@
 
 import pandas as pd
 
+FIND_REPLACE = {  # original relative to /docs : redirect target
+    "<CONTRIBUTING.md>": "[Contributing Section](development/#CONTRIBUTING)",
+    "(CODE_OF_CONDUCT.md)": "(development/#CODE_OF_CONDUCT)",
+    "CONTRIBUTING.md)": "development/#CONTRIBUTING)",
+    "<LICENSE>": "[LICENSE](https://github.com/TIDES-transit/TIDES/blob/main/LICENSE)",
+    "contributors.md)": "development/#contributors)",
+    "architecture.md)": "architecture)",
+    "tables.md": "tables",
+}
+
+def define_env(env):
+    """
+    This is the hook for defining variables, macros and filters
+
+    - variables: the dictionary that contains the environment variables
+    - macro: a decorator function, to declare a macro.
+    """
+
+    @env.macro
+    def include_file(filename: str, start_line: int = 0, end_line: int = None):
+        """
+        Include a file, optionally indicating start_line and end_line.
+
+        Will create redirects if specified in FIND_REPLACE in main.py.
+
+        args:
+            filename: file to include, relative to the top directory of the documentation project.
+            start_line (Optional): if included, will start including the file from this line
+                (indexed to 0)
+            end_line (Optional): if included, will stop including at this line (indexed to 0)
+        """
+        full_filename = os.path.join(env.project_dir, filename)
+        with open(full_filename, "r") as f:
+            lines = f.readlines()
+        line_range = lines[start_line:end_line]
+        content = "".join(line_range)
+
+        _filenamebase = env.page.file.url
+
+        for _find, _replace in FIND_REPLACE.items():
+            if _filenamebase in _replace:
+                _replace = _replace.replace(_filenamebase, "")
+                
+            content = content.replace(_find, _replace)
+        return content
+
+    @env.macro
+    def frictionless_spec(
+        spec_path: str = "**/*.spec.json",
+    ) -> str:
+        """Translate the frictionless .spec file to a markdown table.
+
+        Note: Right now if it finds more than one, it will take the first one.
+
+        Args:
+            spec_path (str, optional): base path of repo. Defaults to two directories
+                up from this file.
+
+        Returns: a markdown table string
+        """
+
+        spec_file = glob.glob(spec_path, recursive=True)[0]
+
+        print("Documenting spec_file: ", spec_file)
+
+        # Generate a table for overall file requirements
+        spec_df = read_config(spec_file)
+        spec_df = spec_df.drop(
+            columns=["fullpath", "fullpath_schema", "path", "schema", "name"]
+        ).reset_index()
+        spec_df["name"] = spec_df["name"].apply(
+            lambda x: "[`{}`](tables.md#{})".format(x, x)
```

Comment:
Need to substitute '-' for '_' in x here to make link anchors work.

Reply:
Or configure separator to "_" for the top [markdown extension](https://www.mkdocs.org/user-guide/configuration/#markdown_extensions)